### PR TITLE
Tpoehlman add label action

### PR DESCRIPTION
**Pull request by tpoehlman**
Description: A new helper action that will assign a label to "Stale" and "Late Review" to the PRs in the provided repo

Tests: Unit tests in created test file

### CODE REVIEW

File: src/helpers/add-pr-stale-label.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+export class AddPrStaleLabel extends HelperInputs {
+  owner = '';
+  repo = '';
+}
+
+/* Adds labels to Pull Requests when they become "Stale" or "Late Review"
+ * @param [Array] prs: array of the pull request ids
+ * @param [string] owner: name of the repo's owner
+ * @param [string] repo: name of the repo
+ */
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:31]: No JS docs please

File: src/helpers/add-pr-stale-label.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+export const addPrStaleLabel = async ({ owner, repo }: AddPrStaleLabel) => {
+  // Get all pull requests
+  var pull_requests = await octokit.pulls.list({
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:34]: No `var` please

File: src/helpers/add-pr-stale-label.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    owner: owner,
+    repo: repo,
+    state: "open"
+  }) as ListResponseType;
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:38]: Do you need this typecast here?

File: src/helpers/add-pr-stale-label.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  var pull_request_data;
+  var pr;
+  var num_requests = pull_requests.data.length;
+  var label;
+
+  // Loop through all of the issue numbers
+  for (var i = 0; i < num_requests; i++) {
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:46]: We use [`Bluebird.map()`](https://www.npmjs.com/package/bluebird) in this repo for multiple async operations

File: src/helpers/add-pr-stale-label.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  }
+};
+
+/* Checks if the given label type is needed for the Pull Request prData
+ * @param [PullResponseType] prData: data object returned by octokit.pulls.get()
+ * @param [string] label: Label wanted to be added (currently supporting labels STALE and LATE_REVIEW)
+ * @return [Boolean] whether or not the provided label is needed for the pull request
+ */
+const isLabelNeeded = async ( prData: PullResponseType, label: string )=> {
+  if (prData == null) {
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:81]:
```suggestion
  if (!prData) {
```

File: src/helpers/add-pr-stale-label.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    return false;
+  }
+  const last_updated = new Date(prData.data.updated_at);
+  const now = new Date();
+  const age = now.getTime() - last_updated.getTime();
+  const margeable_state = prData.data.mergeable_state;
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:87]: Use object desctructuring
```suggestion
  const { mergeable_state, updated_at } = prData.data;
```

File: src/helpers/add-pr-stale-label.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      state = "blocked";
+      break;
+    default:
+      return false;
+  }
+
+  if ( age > 86400000 * days && margeable_state == state ) {
+    return true;
+  }
+  return false;
```
Comment (danadajian) [inline src/helpers/add-pr-stale-label.ts:106]:
```suggestion
 return age > 86400000 * days && margeable_state == state
```

**Review by danadajian — COMMENTED** [review]
Left some initial comments. The stale piece can be accomplished using [actions/stale](https://github.com/actions/stale)

File: src/helpers/add-pr-late-label.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+const labelPullRequest = async ( pull_request: any, owner: string, repo: string) => {
+  const pr = parseInt(pull_request.id);
+  // Get the PR
+  const pull_request_data = await octokit.pulls.get({
+      owner: owner,
+      repo: repo,
+      pull_number: pr
+    }) as PullResponseType;
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:56]: I don't think this typecast is necessary. Typescript should be able to infer the type here!

File: src/helpers/add-pr-late-label.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  const last_updated = new Date( updated_at );
+  const now = new Date();
+  const age = now.getTime() - last_updated.getTime();
+
+  const days = 2;
+  const state = "blocked";
+
+  return age > 86400000 * days && mergeable_state == state;
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:84]: Let's clarify what `86400000` is with a good variable name!

Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:84]: I think we can inline the 'state' here 
```suggestion
  return age > 86400000 * days && mergeable_state == 'blocked';
```

File: src/helpers/add-pr-late-label.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  octokit.issues.addLabels({
+    labels: [LATE_REVIEW],
+    issue_number: pr,
+    owner: owner,
+    repo: repo
+  });
+}
+
+const isLabelNeeded = async ( { mergeable_state, updated_at}: any ) => {
+  if (!mergeable_state || !updated_at) {
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:73]: I'm not sure it's possible for either of these to be falsy values from github...

File: src/helpers/add-pr-late-label.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  const pr = parseInt(pull_request.id);
+  // Get the PR
+  const pull_request_data = await octokit.pulls.get({
+      owner: owner,
+      repo: repo,
+      pull_number: pr
+    }) as PullResponseType;
+
+  // Checks if the PR is within the Late Review timeframe
+  if (!pull_request_data || !await isLabelNeeded(pull_request_data.data)) {
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:59]: `pull_request_data` will never be falsy. You can always assume `pull_request.data.data` is defined

File: test/helpers/add-pr-late-label.test.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
+/*
```
Comment (danadajian) [inline test/helpers/add-pr-late-label.test.ts:1]: Check out [this example](https://github.com/ExpediaGroup/github-helpers/blob/main/test/helpers/remove-pr-from-merge-queue.test.ts#L36) for how to mock the `Date` object

File: src/helpers/add-pr-late-label.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+export class AddPrLateReviewLabels extends HelperInputs {
+  owner = '';
+  repo = '';
+}
+
+export const addPrLateReviewLabels = async ({ owner, repo }: AddPrLateReviewLabels) => {
+  var page = 1;
+  var pull_requests = await getPRs(owner, repo, page) as ListResponseType;
+
+  while (pull_requests.data.length > 0 ) {
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:32]: Check out [this example](https://github.com/ExpediaGroup/github-helpers/blob/main/src/helpers/manage-merge-queue.ts#L90) of paginating over all open pull requests! I think we could extract that into a shared util function and use it here

File: src/helpers/add-pr-late-label.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    sort: "created",
+    per_page: 100,
+    page: page
+  }) as ListResponseType;
+}
+
+const labelPullRequest = async ( pull_request: any, owner: string, repo: string) => {
+  const pr = parseInt(pull_request.id);
+  // Get the PR
+  const pull_request_data = await octokit.pulls.get({
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:69]: We should already have the necessary PR data from our initial `pulls.list` call

File: src/helpers/add-pr-late-label.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    repo: repo
+  });
+}
+
+const isLabelNeeded = async ( { mergeable_state, updated_at}: any ) => {
+  const last_updated = new Date( updated_at );
+  const now = new Date();
+  const age = now.getTime() - last_updated.getTime();
+  const oneDay = 86400000;
+  return age > oneDay * 2 && mergeable_state == "blocked";
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:94]: We should make the number of days configurable as an input to this action and make it required

File: test/helpers/add-pr-late-label.test.ts [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        })
+      }
+    }
+  }))
+}));
+
+jest.spyOn(Date, 'now').mockImplementation(() => new Date('2022-08-04T10:00:00Z').getTime());
+
+describe('addPrLateReviewLabels', () => {
+  jest.setTimeout(9999999);
```
Comment (danadajian) [inline test/helpers/add-pr-late-label.test.ts:68]: not needed

File: test/helpers/add-pr-late-label.test.ts [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+limitations under the License.
+*/
+
+import { LATE_REVIEW } from '../../src/constants';
+import { addPrLateReviewLabels } from '../../src/helpers/add-pr-late-label';
+import { octokit } from '../../src/octokit';
+
+jest.mock('@actions/core');
+jest.mock('@actions/github', () => ({
+  getOctokit: jest.fn(() => ({
```
Comment (danadajian) [inline test/helpers/add-pr-late-label.test.ts:20]: Nitpick, but can we put the mocks in the actual test block where they are needed instead of all at the top of the file? It's hard to follow what the mocked return values are for each test

File: src/helpers/add-pr-late-label.ts (hunk 14 above)
Comment (tpoehlman) [reply src/helpers/add-pr-late-label.ts:69]: I wasnt able to find mergeable as a parameter returned by the list 
https://docs.github.com/en/rest/pulls/pulls#list-pull-requests
, it _does_ have the updated_at field however, so if we can find a way to get the mergeable, or find an alternative way to determine status, we can do away with the call on ilne 69

Comment (danadajian) [reply src/helpers/add-pr-late-label.ts:69]: Looks like we can actually use the `requested_reviewers` and `requested_teams` fields on each listed PR to determine if there are open requests out for a review. So we shouldn't need to make this additional call

File: src/helpers/add-pr-late-label.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    owner: owner,
+    repo: repo,
+    state: "open",
+    sort: "created",
+    per_page: 100,
+    page: page
+  }) as ListResponseType;
+}
+
+const labelPullRequest = async ( pull_request: any, owner: string, repo: string) => {
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:66]: Try not to use `any`. See if you can get to the point where `npm run lint` and `npm run package` work

File: src/helpers/add-pr-late-label.ts [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  });
+}
+
+const isLabelNeeded = async ( { requested_reviewers, requested_teams, updated_at}: any, numDays: number ) => {
+  const last_updated = new Date( updated_at );
+  const now = new Date();
+  const age = now.getTime() - last_updated.getTime();
+  const oneDay = 86400000;
+  // No response over the given number of days & is waiting on reviewers
+  return age > oneDay * numDays && (requested_reviewers || requested_teams);
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:89]: Nice!

File: test/helpers/add-pr-late-label.test.ts (hunk 17 above)
Comment (tpoehlman) [reply test/helpers/add-pr-late-label.test.ts]: Moved what we could, for some reason (although docs say it should be fine) if the original jest.mock's are not above the import it does not work

File: src/helpers/add-pr-late-label.ts (hunk 18 above)
Comment (tpoehlman) [reply src/helpers/add-pr-late-label.ts:65]: Should I create a custom type for the data? the datatype of the return from octokit is "ReastEndpointMethodTypes["pulls"]["list"]["response"], since its built in octokits weird automated type structure.  Not sure if there is a clean way to type its ".data" field

File: src/helpers/add-pr-late-label.ts (hunk 13 above)
Comment (tpoehlman) [reply src/helpers/add-pr-late-label.ts:31]: Would you like me to implement this, or leave it a note for a future improvement?

Comment (danadajian) [reply src/helpers/add-pr-late-label.ts:31]: I opened [this PR](https://github.com/tpoehlman/github-helpers/pull/1) to your fork

**Comment by danadajian** [comment]
@tpoehlman I opened https://github.com/tpoehlman/github-helpers/pull/1 to your fork which should hopefully help. You'll just need to get the test to pass

File: src/utils/paginate-open-pull-requests.ts [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import { PullRequestList } from '../types/github';
```
Comment (danadajian) [inline src/utils/paginate-open-pull-requests.ts:1]: I forgot to add a copyright header to this file

File: src/helpers/add-pr-late-label.ts [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+  const openPullRequests = await paginateAllOpenPullRequests();
+
+  return map(openPullRequests, pr => {
+    if (!isLabelNeeded(pr, Number(days))) {
+      return;
+    }
+
+    return octokit.issues.addLabels({
+      labels: [LATE_REVIEW],
+      issue_number: pr.id,
```
Comment (danadajian) [inline src/helpers/add-pr-late-label.ts:37]: I assume you want the pull request number here, no? I think id is something totally different

File: test/helpers/add-pr-late-label.test.ts [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+jest.mock('../../src/octokit', () => ({
+  octokit: {
+    issues: {
+      addLabels: jest.fn()
+    },
+    pulls: {
+      list: mockList
+    }
+  }
+}));
```
Comment (danadajian) [inline test/helpers/add-pr-late-label.test.ts:24]: Imports should always come first in a file

Comment (tpoehlman) [reply test/helpers/add-pr-late-label.test.ts:24]: For some reason, when running with these specific jests after the imports, it doesnt seem to work at all.  Only way I could find to get it to work was to put it above? I think it has something to do with the difference between mocking '@action/octokit' and mocking the actual file '../../src/octokit'

**Comment by danadajian** [comment]
@tpoehlman I cleaned up the test file and renamed the helper to be a bit more clear. The last thing remaining is to add documentation on the README - could you do that?

File: .github/workflows/add-late-review-label.yml [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.event.pull_request.head.ref }}
+          repository: ${{ github.event.pull_request.head.repo.full_name }}
+          token: ${{ secrets.GH_PERSONAL_ACCESS_TOKEN }}
+
+      - uses: ./
+        with:
+          helper: add-pr-approval-label
```
Comment (danadajian) [inline .github/workflows/add-late-review-label.yml:23]: I think you meant to use your helper!

File: README.md [hunk 24]
```diff
@@ -43,6 +43,9 @@ Each of the following helpers are defined in a file of the same name in `src/hel
 ### [add-labels](.github/workflows/add-labels.yml)
   * Adds one or more labels to a PR
 
+### [add-late-review-label](.github/workflows/add-late-review-label.yml)
+  * Adds a `LATE REVIEW` label to all PRs that are open and waiting for review for over 2 days
```
Comment (danadajian) [inline README.md:47]:
```suggestion
  * Adds a `LATE REVIEW` label to all PRs that are open and waiting for review for over a certain number of days
```

**Review by danadajian — APPROVED** [review]
